staging/src/k8s.io/apiserver/pkg/endpoints/filters/read_write_deadline_test.go

```diff
@@ -0,0 +1,1459 @@
+		clientErrVerifier  verifier
+		handlerErrVerifier verifier
+	}{
+		{
+			protoMajor: 1,
+			// write timeout is set to 100ms, a wait of 5s should
+			// be enough to withstand flakes in CI.
+			waiter:             &waitWithDuration{after: 5 * time.Second},
+			clientErrVerifier:  wantContains{contains: "local error: tls: bad record MAC"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+		{
+			protoMajor:         2,
+			waiter:             &waitWithChannelClose{after: make(chan time.Time)},
+			clientErrVerifier:  wantContains{"stream error: stream ID 1; INTERNAL_ERROR; received from peer"},
+			handlerErrVerifier: wantError{err: os.ErrDeadlineExceeded},
+		},
+	}
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("HTTP/%d", test.protoMajor), func(t *testing.T) {
+			handlerErrCh := make(chanErr, 1)
+			server := httptest.NewUnstartedServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+				defer close(handlerErrCh)
+				if req.ProtoMajor != test.protoMajor {
+					t.Errorf("server: expected ProtoMajor: %d, but got: %s", test.protoMajor, req.Proto)
+					return
+				}
+				flusher, ok := w.(interface{ FlushError() error })
+				if !ok {
+					t.Errorf("server: expected ResponseWriter object to implement FlushError")
+					return
+				}
+
+				// a) set write deadline for this request
+				ctrl := http.NewResponseController(w)
+				if err := ctrl.SetWriteDeadline(time.Now().Add(deadline)); err != nil {
+					t.Errorf("server: unexpected error from SetWriteDeadline: %v", err)
+					return
+				}
+
+				// b) wait until the write deadline exceeds
+				<-test.waiter.wait()
+
+				// c) write a few bytes so there is something in the buffer to
+				// flush, the client should never see these bytes in the response.
+				if _, err := w.Write([]byte("hello")); err != nil {
+					t.Errorf("server: unexpected error from Write after timeout: %v", err)
+					return
+				}
+				// d) flush the payload in the buffer
+				// NOTE: FlushError, on the other hand, is expected
+				// to return a timeout error immediately.
+				if err := flusher.FlushError(); err != nil {
+					handlerErrCh <- err
+				}
+
+				// NOTE: the handler goroutine is expected to terminate as soon
+				// as it receives an error from either Write or FlushError. This
+				// is how we prevent a request handler from running indefinitely.
+			}))
+
+			defer server.Close()
+			if test.protoMajor == 2 {
+				server.EnableHTTP2 = true
+			}
+			server.StartTLS()
+
+			func() {
+				defer test.waiter.close()
+
+				_, err := server.Client().Get(server.URL)
+				// e) verify that the client receives the appropriate error
+				test.clientErrVerifier.verify(t, err)
+			}()
+
+			// f) verify that the handler returns the right error
+			err := handlerErrCh.wait(t, time.After(wait.ForeverTestTimeout))
+			test.handlerErrVerifier.verify(t, err)
+		})
+	}
+}
+
+func TestPerRequestWriteDeadlineWithTimeoutOccursAfterHandlerWrites(t *testing.T) {
```

Comment:
TestFlushErrorAfterRequestDeadline

In general, don't describe just the preconditions, but describe the claimed behaviour in test names.

Reply:
renamed with claimed behavior, I also want to keep the prefix `TestPerRequestWith` so it is easier to run these tests in a group.